### PR TITLE
Upsert does not use ttl value

### PR DESCRIPTION
**Pull request by dadoonet**
When running an upsert which defines a ttl value, the ttl value is set to null and ignored.

Related to #3256#issuecomment-64963409

### CODE REVIEW

**Comment by TheDeveloper** [comment]
I am running into this issue also when the TTL is set as a default on the index, not on the update request.

Is this patch likely to get merged soon?

Am I right in saying the workaround for now is to ensure the new TTL is set via the script's `_ttl` global?

**Comment by dadoonet** [comment]
@s1monw I rebased and added a new commit to move from integration tests to unit tests.
Let me know what you think.

File: src/main/java/org/elasticsearch/action/update/UpdateHelper.java [hunk 1]
```diff
@@ -79,35 +79,48 @@ public Result prepare(UpdateRequest request, IndexShard indexShard) {
...
+                        // we need to unwrap the ctx...
+                        ctx = (Map<String, Object>) script.unwrap(ctx);
+                    }
                 } catch (Exception e) {
                     throw new ElasticsearchIllegalArgumentException("failed to execute script", e);
-                }                
+                }
                 //Allow the script to set TTL using ctx._ttl
-                ttl = getTTLFromScriptContext(ctx);
+                if (ttl < 0) {
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/action/update/UpdateHelper.java]: can ttl be null here?

File: src/main/java/org/elasticsearch/action/update/UpdateHelper.java [hunk 2]
```diff
@@ -79,35 +79,48 @@ public Result prepare(UpdateRequest request, IndexShard indexShard) {
...
                 new String[]{RoutingFieldMapper.NAME, ParentFieldMapper.NAME, TTLFieldMapper.NAME, TimestampFieldMapper.NAME},
                 true, request.version(), request.versionType(), FetchSourceContext.FETCH_SOURCE, false);
+        return prepare(request, getResult);
+    }
 
+    /**
+     * Prepares an update request by converting it into an index or delete request or an update response (no action).
+     */
+    @SuppressWarnings("unchecked")
+    public Result prepare(UpdateRequest request, final GetResult getResult) {
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/action/update/UpdateHelper.java]: does this need to be public or can it be package private?

**Comment by s1monw** [comment]
left minor comments looks good in general

**Comment by dadoonet** [comment]
@s1monw I rebased and updated the code based on your comments. Let me know if I can merge this in.

File: core/src/main/java/org/elasticsearch/action/update/UpdateHelper.java [hunk 3]
```diff
@@ -73,39 +73,51 @@ public UpdateHelper(Settings settings, ScriptService scriptService) {
...
                 // Tell the script that this is a create and not an update
                 ctx.put("op", "create");
                 ctx.put("_source", upsertDoc);
                 try {
-                    ExecutableScript script = scriptService.executable(request.script, ScriptContext.Standard.UPDATE);
-                    script.setNextVar("ctx", ctx);
-                    script.run();
-                    // we need to unwrap the ctx...
-                    ctx = (Map<String, Object>) script.unwrap(ctx);
+                    if (scriptService != null) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/update/UpdateHelper.java]: you could extract this in a private method to share some code with the scritp execution below.

**Comment by s1monw** [comment]
left a cosmetic comment. Feel free to push once fixed... in other words LGTM

**Comment by dadoonet** [comment]
@s1monw I updated the code because tests were failing with my PR.

I changed 2 lines:

* change `long` to `Long`: https://github.com/dadoonet/elasticsearch/blob/b4aa472716fa2dec703591fd115cfcd607014872/core/src/main/java/org/elasticsearch/action/update/UpdateHelper.java#L93
* if (ttl < 0) we need to pass `null` to the IndexRequest https://github.com/dadoonet/elasticsearch/blob/b4aa472716fa2dec703591fd115cfcd607014872/core/src/main/java/org/elasticsearch/action/update/UpdateHelper.java#L129

If we don't do it, it's rejected by https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/index/IndexRequest.java#L327-329

Do you agree with those last changes? All tests pass on my laptop now.

**Comment by s1monw** [comment]
> change long to Long: https://github.com/dadoonet/elasticsearch/blob/b4aa472716fa2dec703591fd115cfcd607014872/core/src/main/java/org/elasticsearch/action/update/UpdateHelper.java#L93

why this change?

**Comment by dadoonet** [comment]
Because this line: https://github.com/dadoonet/elasticsearch/blob/b4aa472716fa2dec703591fd115cfcd607014872/core/src/main/java/org/elasticsearch/action/update/UpdateHelper.java#L105 was making a NPE. You can't set `null` to a `long`

**Comment by s1monw** [comment]
fair enough.... LGTM

**Pull request merged**